File: section30#Huofuman/C++/Hough.cpp
#include <iostream>
#include <string>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

static bool houghTest(Mat imgOri);

int main()
{
	Mat imgOri = imread("building.jpg");
	if (imgOri.empty())
	{
		cout << "Cannot read image" << endl;
	}

	houghTest(imgOri);

	waitKey(0);
	return true;
}

static bool houghTest(Mat imgOri)
{
	Mat imgCan;
	imshow("imgOri", imgOri);

	//GaussianBlur(imgOri, imgOri, Size(7, 7), 0);

	Canny(imgOri, imgCan, 100, 160, 3, false);
	imshow("Canny", imgCan);

	vector<Vec2f> lines;
	Mat imgHough = imgOri.clone();
	Mat imgHoughP = imgOri.clone();

	HoughLines(imgCan, lines, 1, CV_PI / 180, 200, 0, 0);

	Point pt1, pt2;
	for (size_t i = 0; i < lines.size(); i++)
	{
		float rho = lines[i][0];
		float theta = lines[i][1];
		double a = cos(theta), b = sin(theta);
		double x0 = a * rho, y0 = b * rho;
		pt1.x = cvRound(x0 + 1000 * (-b));
		pt1.y = cvRound(y0 + 1000 * (a));
		pt2.x = cvRound(x0 - 1000 * (-b));
		pt2.y = cvRound(y0 - 1000 * (a));
		line(imgHough, pt1, pt2, Scalar(0, 0, 255), 3, LINE_AA);
	}

	imshow("imgHough", imgHough);


	vector<Vec4i> Plines;
	HoughLinesP(imgCan, Plines, 1, CV_PI / 180, 80, 30, 10);
	for (size_t i = 0; i < Plines.size(); i++)
	{
		line(imgHoughP, Point(Plines[i][0], Plines[i][1]),
			Point(Plines[i][2], Plines[i][3]), Scalar(0, 255, 0), 1, 8);
	}

	imshow("imgHoughP", imgHoughP);


	return true;
}

File: section30#Huofuman/Python/Hough.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# author：lowkeyway time:9/17/2019

import sys
import cv2 as cv
import numpy as np
import matplotlib.pyplot as plt

def houghTest(imgOri):
    cv.imshow("imgOri", imgOri)
    imgCan = cv.Canny(imgOri, 100, 200)
    cv.imshow("imgCan", imgCan)

    imgHough = imgOri.copy()
    imgHoughP = imgOri.copy()
    Hlines = cv.HoughLines(imgCan, 1, np.pi/180, 150)

    for line in Hlines:
        rho = line[0][0]
        theta = line[0][1]
        x0 = rho*np.cos(theta)
        y0 = rho*np.sin(theta)

        p1 = (int(x0 + 1000*np.sin(theta)), int(y0 - 1000*np.cos(theta)))
        p2 = (int(x0 - 1000*np.sin(theta)), int(y0 + 1000*np.cos(theta)))

        cv.line(imgHough, p1, p2, (0, 0, 255), 3, cv.LINE_AA)

    HPlines = cv.HoughLinesP(imgCan, 1, np.pi/180, 50, None, 50, 10)
    for line in HPlines:
        x0, y0, x1, y1 = line[0]
        cv.line(imgHoughP, (x0, y0), (x1, y1), (0, 255, 0), 1, cv.LINE_AA)

    cv.imshow("imgHough", imgHough)
    cv.imshow("imgHoughP", imgHoughP)


    plt.figure("HoughPic")
    plt.subplot(221); plt.imshow(imgOri); plt.title("ImgOri")
    plt.subplot(222); plt.imshow(imgCan, "gray"); plt.title("imgCan")
    plt.subplot(223); plt.imshow(imgHough, "gray"); plt.title("imgHough")
    plt.subplot(224); plt.imshow(imgHoughP, "gray"); plt.title("imgHoughP")

    plt.show()


def main_func(argv):
    imgOri = cv.imread("sudoku.png")
    if imgOri is None:
        print("Cannot load this picture!")
        return

    houghTest(imgOri)
    cv.waitKey(0)

if __name__ == '__main__':
    main_func(sys.argv)